File: app/src/test/java/com/coinninja/coinkeeper/ui/actionbar/ActionBarControllerTest.java
package com.coinninja.coinkeeper.ui.actionbar;

import android.util.TypedValue;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import com.coinninja.coinkeeper.R;
import com.coinninja.coinkeeper.ui.actionbar.managers.TitleViewManager;
import com.coinninja.coinkeeper.ui.base.MenuItemClickListener;

import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.anyBoolean;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


@RunWith(MockitoJUnitRunner.class)
public class ActionBarControllerTest {

    @Mock
    private TitleViewManager titleViewManager;

    @Mock
    private AppCompatActivity context;

    @InjectMocks
    private ActionBarController controller;

    @After
    public void tearDown() throws Exception {
        titleViewManager = null;
        context = null;
        controller = null;
    }

    @Test
    public void actionbar_gone_configuration_test() {

        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = R.id.actionbar_gone;

        controller.setTheme(context, actionBarTyped);

        assertTrue(controller.isActionBarGone);
        assertFalse(controller.isTitleUppercase);
        assertNull(controller.isUpEnabled);
        assertNull(controller.optionMenuLayout);
    }


    @Test
    public void actionbar_light_up_on_configuration_test() {
        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = R.id.actionbar_up_on;

        controller.setTheme(context, actionBarTyped);

        assertNull(controller.isActionBarGone);
        assertTrue(controller.isUpEnabled);
        assertFalse(controller.isTitleUppercase);
        assertNull(controller.optionMenuLayout);
    }

    @Test
    public void actionbar_light_up_off_configuration_test() {
        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = R.id.actionbar_up_off;

        controller.setTheme(context, actionBarTyped);

        assertNull(controller.isActionBarGone);
        assertFalse(controller.isUpEnabled);
        assertFalse(controller.isTitleUppercase);
        assertNull(controller.optionMenuLayout);
    }

    @Test
    public void actionbar_light_up_off_skip_on_configuration_test() {

        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = R.id.actionbar_up_off_skip_on;

        controller.setTheme(context, actionBarTyped);

        assertNull(controller.isActionBarGone);
        assertFalse(controller.isUpEnabled);
        assertFalse(controller.isTitleUppercase);
        assertThat(controller.optionMenuLayout, equalTo(R.menu.actionbar_light_skip_menu));
    }

    @Test
    public void actionbar_light_up_on_skip_on_configuration_test() {

        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = R.id.actionbar_up_on_skip_on;

        controller.setTheme(context, actionBarTyped);

        assertNull(controller.isActionBarGone);
        assertTrue(controller.isUpEnabled);
        assertFalse(controller.isTitleUppercase);
        assertThat(controller.optionMenuLayout, equalTo(R.menu.actionbar_light_skip_menu));
    }

    @Test(expected = IllegalStateException.class)
    public void throw_illegal_state_exception_when_theme_unknown() {

        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = -1;

        controller.setTheme(context, actionBarTyped);

    }

    @Test
    public void remove_can_container_layout_if_them_is_action_gone() {
        View cnContainerLayout = mock(View.class);
        when((context).findViewById(R.id.cn_appbar_layout_container)).thenReturn(cnContainerLayout);
        controller.isActionBarGone = true;

        controller.displayTitle(context);

        verify(cnContainerLayout).setVisibility(View.GONE);
    }

    @Test
    public void init_title_view_directly_after_setting_up_theme() {
        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = R.id.actionbar_up_on_skip_on;
        ActionBar supportActionBar = mock(ActionBar.class);
        TextView titleView = mock(TextView.class);
        when(context.getSupportActionBar()).thenReturn(supportActionBar);
        when(context.findViewById(R.id.appbar_title)).thenReturn(titleView);


        controller.setTheme(context, actionBarTyped);


        verify(titleViewManager).setActionBar(supportActionBar);
        verify(titleViewManager).setTitleView(titleView);
    }

    @Test
    public void do_not_init_title_view_when_theme_is_action_gone() {
        TypedValue actionBarTyped = new TypedValue();
        actionBarTyped.resourceId = R.id.actionbar_gone;
        ActionBar supportActionBar = mock(ActionBar.class);
        TextView titleView = mock(TextView.class);

        controller.setTheme(context, actionBarTyped);

        verify(titleViewManager, never()).setActionBar(supportActionBar);
        verify(titleViewManager, never()).setTitleView(titleView);
    }

    @Test
    public void if_upper_case_is_true_render_uppercase_title_view_when_displayTitle_is_called() {
        controller.titleViewManager = titleViewManager;
        controller.isTitleUppercase = true;

        controller.displayTitle(context);

        verify(titleViewManager).renderUpperCaseTitleView();
    }

    @Test
    public void if_up_is_enabled_then_setDisplayHomeAsUpEnabled_true() {
        Menu menu = mock(Menu.class);
        ActionBar supportActionBar = mock(ActionBar.class);
        when(context.getSupportActionBar()).thenReturn(supportActionBar);
        controller.isUpEnabled = true;

        controller.inflateActionBarMenu(context, menu);

        verify(supportActionBar).setDisplayHomeAsUpEnabled(true);
    }

    @Test
    public void if_up_is_not_enabled_then_setDisplayHomeAsUpEnabled_false() {
        Menu menu = mock(Menu.class);
        ActionBar supportActionBar = mock(ActionBar.class);
        when(context.getSupportActionBar()).thenReturn(supportActionBar);
        controller.isUpEnabled = false;

        controller.inflateActionBarMenu(context, menu);

        verify(supportActionBar).setDisplayHomeAsUpEnabled(false);
    }


    @Test
    public void if_up_is_null_then_do_nothing() {
        Menu menu = mock(Menu.class);
        ActionBar supportActionBar = mock(ActionBar.class);
        controller.isUpEnabled = null;

        controller.inflateActionBarMenu(context, menu);

        verify(supportActionBar, never()).setDisplayHomeAsUpEnabled(anyBoolean());
    }


    @Test
    public void if_optionMenuLayout_has_any_value_inflate_it() {
        Menu menu = mock(Menu.class);
        MenuInflater menuInflater = mock(MenuInflater.class);
        when(context.getMenuInflater()).thenReturn(menuInflater);
        controller.optionMenuLayout = R.menu.actionbar_dark_close_menu;

        controller.inflateActionBarMenu(context, menu);

        verify(menuInflater).inflate(R.menu.actionbar_dark_close_menu, menu);
    }

    @Test
    public void if_optionMenuLayout_is_null_then_do_nothing() {
        Menu menu = mock(Menu.class);
        MenuInflater menuInflater = mock(MenuInflater.class);
        controller.optionMenuLayout = null;

        controller.inflateActionBarMenu(context, menu);

        verify(menuInflater, never()).inflate(R.menu.actionbar_dark_close_menu, menu);
    }

    @Test
    public void if_optionMenuLayout_is_null_then_onMenuItemClicked_do_nothing() {
        MenuItem item = mock(MenuItem.class);
        MenuItemClickListener menuItemClickListener = mock(MenuItemClickListener.class);
        controller.setMenuItemClickListener(menuItemClickListener);
        controller.optionMenuLayout = null;

        boolean itemClicked = controller.onMenuItemClicked(item);

        assertFalse(itemClicked);
        verify(menuItemClickListener, never()).onCloseClicked();
        verify(menuItemClickListener, never()).onSkipClicked();
    }

    @Test
    public void if_action_skip_btn_clicked_then_call_menuItemClickListener_onSkipClicked() {
        MenuItem item = mock(MenuItem.class);
        when(item.getItemId()).thenReturn(R.id.action_skip_btn);
        MenuItemClickListener menuItemClickListener = mock(MenuItemClickListener.class);
        controller.setMenuItemClickListener(menuItemClickListener);
        controller.optionMenuLayout = R.menu.actionbar_light_skip_menu;

        boolean itemClicked = controller.onMenuItemClicked(item);

        verify(menuItemClickListener).onSkipClicked();
        assertTrue(itemClicked);
        verify(menuItemClickListener, never()).onCloseClicked();
    }

    @Test
    public void if_action_close_btn_clicked_then_call_menuItemClickListener_onSkipClicked() {
        MenuItem item = mock(MenuItem.class);
        when(item.getItemId()).thenReturn(R.id.action_close_btn);
        MenuItemClickListener menuItemClickListener = mock(MenuItemClickListener.class);
        controller.setMenuItemClickListener(menuItemClickListener);
        controller.optionMenuLayout = R.menu.actionbar_dark_close_menu;

        boolean itemClicked = controller.onMenuItemClicked(item);

        verify(menuItemClickListener).onCloseClicked();
        assertTrue(itemClicked);
        verify(menuItemClickListener, never()).onSkipClicked();
    }

    @Test
    public void if_a_menu_item_was_clicked_but_is_unknown_the_return_false() {
        MenuItem item = mock(MenuItem.class);
        when(item.getItemId()).thenReturn(-1);
        MenuItemClickListener menuItemClickListener = mock(MenuItemClickListener.class);
        controller.setMenuItemClickListener(menuItemClickListener);
        controller.optionMenuLayout = R.menu.actionbar_dark_close_menu;

        boolean itemClicked = controller.onMenuItemClicked(item);

        assertFalse(itemClicked);
        verify(menuItemClickListener, never()).onCloseClicked();
        verify(menuItemClickListener, never()).onSkipClicked();
    }

    @Test
    public void if_isActionBarGone_false_then_when_updateTitle_then_call_titleViewManager_renderTitleView() {
        controller.isActionBarGone = false;

        controller.updateTitle("--- some new title");

        verify(titleViewManager).renderTitleView("--- some new title");
    }


    @Test
    public void if_isActionBarGone_true_then_do_nothing_when_updateTitle() {
        controller.isActionBarGone = true;

        controller.updateTitle("--- some new title");

        verify(titleViewManager, never()).renderTitleView(anyString());
    }


    @Test
    public void if_isActionBarGone_null_then_do_nothing_when_updateTitle() {
        controller.isActionBarGone = true;

        controller.updateTitle("--- some new title");

        verify(titleViewManager, never()).renderTitleView(anyString());
    }

}

File: app/src/test/java/com/coinninja/coinkeeper/ui/actionbar/ActionBarControllerTest.kt
package com.coinninja.coinkeeper.ui.actionbar

import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.lifecycle.Observer
import com.coinninja.coinkeeper.R
import com.coinninja.coinkeeper.util.DefaultCurrencies
import com.coinninja.coinkeeper.util.currency.BTCCurrency
import com.coinninja.coinkeeper.util.currency.CryptoCurrency
import com.coinninja.coinkeeper.util.currency.FiatCurrency
import com.coinninja.coinkeeper.util.currency.USDCurrency
import com.coinninja.coinkeeper.view.widget.DefaultCurrencyDisplaySyncView
import com.google.android.material.tabs.TabLayout
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockitokotlin2.*
import org.junit.Test
import org.mockito.ArgumentCaptor


class ActionBarControllerTest {

    private val activity: AppCompatActivity = mock()
    private val root: ViewGroup = mock()
    private val charts: ImageButton = mock()
    private val appbarBalance: DefaultCurrencyDisplaySyncView = mock()
    private val appbarBalanceLarge: DefaultCurrencyDisplaySyncView = mock()

    private fun createController(): ActionBarController = ActionBarController(mock()).also {
        whenever(charts.parent).thenReturn(root)
        whenever(appbarBalance.parent).thenReturn(root)
        whenever(appbarBalanceLarge.parent).thenReturn(root)
        whenever(activity.menuInflater).thenReturn(mock())
        whenever(activity.supportActionBar).thenReturn(mock())
        whenever(activity.supportActionBar!!.title).thenReturn("")
        whenever(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)).thenReturn(appbarBalance)
        whenever(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance_large)).thenReturn(appbarBalanceLarge)
        whenever(activity.findViewById<TextView>(R.id.appbar_title)).thenReturn(mock())
        whenever(activity.findViewById<ImageButton>(R.id.appbar_charts)).thenReturn(charts)
        whenever(activity.findViewById<ConstraintLayout>(R.id.cn_content_wrapper)).thenReturn(mock())
        whenever(it.walletViewModel.syncInProgress).thenReturn(mock())
        whenever(it.walletViewModel.chainHoldings).thenReturn(mock())
        whenever(it.walletViewModel.chainHoldingsWorth).thenReturn(mock())
        whenever(it.walletViewModel.defaultCurrencyPreference).thenReturn(mock())
    }

    @Test
    fun actionbar_gone_configuration_test() {
        val controller = createController()
        val actionBarTyped = TypedValue()


        actionBarTyped.resourceId = R.id.actionbar_gone
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isTrue()
        assertThat(controller.isUpEnabled).isFalse()
        assertThat(controller.optionMenuLayout).isNull()
    }

    @Test
    fun actionbar_light__up_on_configuration_test() {
        val controller = createController()
        val actionBarTyped = TypedValue()

        actionBarTyped.resourceId = R.id.actionbar_up_on
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isTrue()
        assertThat(controller.optionMenuLayout).isNull()


        actionBarTyped.resourceId = R.id.actionbar_up_on_balance_on
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isTrue()
        assertThat(controller.optionMenuLayout).isNull()


        actionBarTyped.resourceId = R.id.actionbar_up_on_with_nav_bar
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isTrue()
        assertThat(controller.optionMenuLayout).isNull()

        actionBarTyped.resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isTrue()
        assertThat(controller.optionMenuLayout).isNull()


        actionBarTyped.resourceId = R.id.actionbar_up_on_skip_on
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isTrue()
        assertThat(controller.optionMenuLayout).isEqualTo(R.menu.actionbar_light_skip_menu)
    }

    @Test
    fun actionbar_light__up_off_configuration_test() {
        val controller = createController()
        val actionBarTyped = TypedValue()

        actionBarTyped.resourceId = R.id.actionbar_up_off
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isFalse()
        assertThat(controller.optionMenuLayout).isNull()


        actionBarTyped.resourceId = R.id.actionbar_up_off_close_on
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isFalse()
        assertThat(controller.optionMenuLayout).isEqualTo(R.menu.actionbar_light_close_menu)


        actionBarTyped.resourceId = R.id.actionbar_up_off_skip_on
        controller.setTheme(activity, actionBarTyped)
        assertThat(controller.isActionBarGone).isFalse()
        assertThat(controller.isUpEnabled).isFalse()
        assertThat(controller.optionMenuLayout).isEqualTo(R.menu.actionbar_light_skip_menu)
    }


    @Test(expected = IllegalStateException::class)
    fun throw_illegal_state_exception_when_theme_unknown() {
        val controller = createController()

        val actionBarTyped = TypedValue()
        actionBarTyped.resourceId = -1

        controller.setTheme(activity, actionBarTyped)

    }

    @Test
    fun remove_can_container_layout_if_them_is_action_gone() {
        val controller = createController()
        val cnContainerLayout = mock<View>()
        whenever(activity.findViewById<View>(R.id.appbar_group)).thenReturn(cnContainerLayout)

        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_gone
        }

        controller.setTheme(activity, actionBarTyped)

        verify(cnContainerLayout).visibility = View.GONE
    }

    @Test
    fun if_up_is_enabled_then_setDisplayHomeAsUpEnabled_true() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(activity.supportActionBar)!!.setDisplayHomeAsUpEnabled(true)
    }

    @Test
    fun if_up_is_not_enabled_then_setDisplayHomeAsUpEnabled_false() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_off
        }

        controller.setTheme(activity, actionBarTyped)

        verify(activity.supportActionBar)!!.setDisplayHomeAsUpEnabled(false)
    }

    @Test
    fun set_title_to_app_bar() {
        val controller = createController()
        val title = " --- TITLE --"
        whenever(activity.supportActionBar!!.title).thenReturn(title)
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_off
        }

        controller.setTheme(activity, actionBarTyped)

        val textView = activity.findViewById<TextView>(R.id.appbar_title)!!
        verify(textView).visibility = View.VISIBLE
        verify(textView).text = title
        verify(activity.supportActionBar!!).title = ""
    }

    @Test
    fun set_title_to_app_bar_directly() {
        val controller = createController()
        val title = " --- TITLE --"
        val titleWeDoNotWant = " --- TITLE BAD --"
        whenever(activity.supportActionBar!!.title).thenReturn(titleWeDoNotWant)
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_off
        }
        controller.setTheme(activity, actionBarTyped)


        controller.displayTitle(activity, title)


        val textView = activity.findViewById<TextView>(R.id.appbar_title)!!
        verify(textView, times(2)).visibility = View.VISIBLE
        verify(textView).text = title
        verify(activity.supportActionBar!!, times(2)).title = ""
    }

    @Test
    fun title_bar__gone_when_empty() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_off
        }
        controller.setTheme(activity, actionBarTyped)

        controller.displayTitle(activity, "")

        val textView = activity.findViewById<TextView>(R.id.appbar_title)!!
        verify(textView, atLeastOnce()).visibility = View.GONE
        verify(activity.supportActionBar!!, times(2)).title = ""
    }

    @Test
    fun balance__shows_when_theme_requests_it_up_on_balance_on() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        val balanceView = activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)!!
        verify(balanceView).visibility = View.VISIBLE
    }

    @Test
    fun balance__gone_when_theme_has_no_balance() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(root, atLeastOnce()).removeView(charts)
        verify(root, atLeastOnce()).removeView(appbarBalance)
        verify(root, atLeastOnce()).removeView(appbarBalanceLarge)
    }

    @Test
    fun balance_shows_when_theme_requests_it_drawer_with_balance_on() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        val balanceView = activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)!!
        verify(balanceView).visibility = View.VISIBLE
        verify(root, atLeastOnce()).removeView(charts)
        verify(root, atLeastOnce()).removeView(appbarBalanceLarge)
    }

    @Test
    fun shows_balance_with_charts() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on_charts_on
        }

        controller.setTheme(activity, actionBarTyped)

        val chartsIcon = activity.findViewById<ImageButton>(R.id.appbar_charts)!!
        val balanceView = activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)!!
        verify(chartsIcon).visibility = View.VISIBLE
        verify(balanceView).visibility = View.GONE // HomeScreen will render this
        verify(controller.walletViewModel.syncInProgress, atLeastOnce()).observe(eq(activity), any())
        verify(controller.walletViewModel.chainHoldingsWorth, atLeastOnce()).observe(eq(activity), any())
        verify(controller.walletViewModel.defaultCurrencyPreference, atLeastOnce()).observe(eq(activity), any())
        verify(controller.walletViewModel.chainHoldings, atLeastOnce()).observe(eq(activity), any())
    }

    @Test
    fun balance_loads_data_when_enabled() {
        val controller = createController()
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(controller.walletViewModel).loadCurrencyDefaults()
        verify(controller.walletViewModel).loadHoldingBalances()
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun balance_view_observes_syncing_activity() {
        val controller = createController()
        val argumentCaptor: ArgumentCaptor<Observer<Boolean>> = ArgumentCaptor.forClass(Observer::class.java)
                as ArgumentCaptor<Observer<Boolean>>
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(controller.walletViewModel.syncInProgress, atLeastOnce()).observe(eq(activity), argumentCaptor.capture())

        argumentCaptor.value.onChanged(true)
        verify(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)).showSyncingUI()
        argumentCaptor.value.onChanged(false)
        verify(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)).hideSyncingUI()
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun balance_view_observes_chainHoldings_changes() {
        val controller = createController()
        val holdings = BTCCurrency(1000L)
        val worth = USDCurrency(500.00)
        val currenciesPreferences = DefaultCurrencies(USDCurrency(), BTCCurrency())
        whenever(controller.walletViewModel.chainHoldingsWorth.value).thenReturn(worth)
        whenever(controller.walletViewModel.defaultCurrencyPreference.value).thenReturn(currenciesPreferences)
        val argumentCaptor: ArgumentCaptor<Observer<CryptoCurrency>> = ArgumentCaptor.forClass(Observer::class.java)
                as ArgumentCaptor<Observer<CryptoCurrency>>
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(controller.walletViewModel.chainHoldings, atLeastOnce()).observe(eq(activity), argumentCaptor.capture())

        argumentCaptor.value.onChanged(holdings)

        verify(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)).renderValues(currenciesPreferences, holdings, worth)
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun balance_view_observes_chainHoldingsWorth_changes() {
        val controller = createController()
        val holdings = BTCCurrency(1000L)
        val worth = USDCurrency(500.00)
        val currenciesPreferences = DefaultCurrencies(USDCurrency(), BTCCurrency())
        whenever(controller.walletViewModel.chainHoldings.value).thenReturn(holdings)
        whenever(controller.walletViewModel.defaultCurrencyPreference.value).thenReturn(currenciesPreferences)
        val argumentCaptor: ArgumentCaptor<Observer<FiatCurrency>> = ArgumentCaptor.forClass(Observer::class.java)
                as ArgumentCaptor<Observer<FiatCurrency>>
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(controller.walletViewModel.chainHoldingsWorth, atLeastOnce()).observe(eq(activity), argumentCaptor.capture())

        argumentCaptor.value.onChanged(worth)
        verify(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)).renderValues(currenciesPreferences, holdings, worth)
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun observes_balance_toggles_default_currency_preference() {
        val controller = createController()
        val holdings = BTCCurrency(1000L)
        val worth = USDCurrency(500.00)
        val currenciesPreferences = DefaultCurrencies(USDCurrency(), BTCCurrency())
        whenever(controller.walletViewModel.chainHoldings.value).thenReturn(holdings)
        whenever(controller.walletViewModel.chainHoldingsWorth.value).thenReturn(worth)
        val argumentCaptor: ArgumentCaptor<Observer<DefaultCurrencies>> = ArgumentCaptor.forClass(Observer::class.java)
                as ArgumentCaptor<Observer<DefaultCurrencies>>
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(controller.walletViewModel.defaultCurrencyPreference, atLeastOnce()).observe(eq(activity), argumentCaptor.capture())

        argumentCaptor.value.onChanged(currenciesPreferences)
        verify(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)).renderValues(currenciesPreferences, holdings, worth)
    }

    @Test
    fun clicking_balance_toggles_default_currency_preference() {
        val controller = createController()
        val argumentCaptor: ArgumentCaptor<View.OnClickListener> = ArgumentCaptor.forClass(View.OnClickListener::class.java)
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)
        verify(activity.findViewById<DefaultCurrencyDisplaySyncView>(R.id.appbar_balance)).setOnClickListener(argumentCaptor.capture())
        argumentCaptor.value.onClick(mock())

        verify(controller.walletViewModel).toggleDefaultCurrencyPreference()
    }

    @Test
    fun hides_tabs_by_default() {
        val controller = createController()
        val tabLayout: TabLayout = mock()
        whenever(activity.findViewById<TabLayout>(R.id.appbar_tabs)).thenReturn(tabLayout)
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }

        controller.setTheme(activity, actionBarTyped)

        verify(tabLayout).visibility = View.GONE
    }

    @Test
    fun adds_tabs_to_tab_view() {
        val controller = createController()
        val tabLayout: TabLayout = mock()
        whenever(activity.findViewById<TabLayout>(R.id.appbar_tabs)).thenReturn(tabLayout)
        val tab = mock<TabLayout.Tab>()
        whenever(tabLayout.getTabAt(0)).thenReturn(tab)
        val actionBarTyped = TypedValue().apply {
            resourceId = R.id.actionbar_up_on_with_nav_bar_balance_on
        }
        controller.setTheme(activity, actionBarTyped)

        controller.addTab(activity, R.layout.home_appbar_tab_1, 0)

        verify(tab).setCustomView(R.layout.home_appbar_tab_1)
    }
}